src/FindIntrinsics.cpp

```diff
@@ -0,0 +1,775 @@
+             narrow_a.type().is_float() == narrow_b.type().is_float())) {
+            Expr result = widening_mul(narrow_a, narrow_b);
+            if (result.type() != op->type) {
+                result = Cast::make(op->type, result);
+            }
+            return mutate(result);
+        }
+
+        if (a.same_as(op->a) && b.same_as(op->b)) {
+            return op;
+        } else {
+            return Mul::make(a, b);
+        }
+    }
+
+    Expr visit(const Div *op) override {
+        if (!find_intrinsics_for_type(op->type)) {
+            return IRMutator::visit(op);
+        }
+
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+
+        if (!op->type.is_float()) {
+            return mutate(lower_int_uint_div(a, b));
+        }
+
+        if (a.same_as(op->a) && b.same_as(op->b)) {
+            return op;
+        } else {
+            return Div::make(a, b);
+        }
+    }
+
+    Expr visit(const Mod *op) override {
+        if (!find_intrinsics_for_type(op->type)) {
+            return IRMutator::visit(op);
+        }
+
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+
+        if (!op->type.is_float()) {
+            return mutate(lower_int_uint_mod(a, b));
+        }
+
+        if (a.same_as(op->a) && b.same_as(op->b)) {
+            return op;
+        } else {
+            return Mod::make(a, b);
+        }
+    }
+
+    template<class MinOrMax>
+    Expr visit_min_or_max(const MinOrMax *op) {
+        if (!find_intrinsics_for_type(op->type)) {
+            return IRMutator::visit(op);
+        }
+
+        Expr a = mutate(op->a);
+        Expr b = mutate(op->b);
+
+        if (const Cast *cast_a = a.as<Cast>()) {
+            Expr cast_b = lossless_cast(cast_a->value.type(), b);
+            if (cast_a->type.can_represent(cast_a->value.type()) && cast_b.defined()) {
+                // This cast can be moved outside the min.
+                return mutate(Cast::make(cast_a->type, MinOrMax::make(cast_a->value, cast_b)));
+            }
+        }
+        if (a.same_as(op->a) && b.same_as(op->b)) {
+            return op;
+        } else {
+            return MinOrMax::make(a, b);
+        }
+    }
+
+    Expr visit(const Min *op) override {
+        return visit_min_or_max(op);
+    }
+
+    Expr visit(const Max *op) override {
+        return visit_min_or_max(op);
+    }
+
+    Expr visit(const Cast *op) override {
+        if (!find_intrinsics_for_type(op->type)) {
+            return IRMutator::visit(op);
+        }
+
+        Expr value = mutate(op->value);
+
+        // This mutator can generate redundant casts. We can't use the simplifier because it
+        // undoes some of the intrinsic lowering here, and it causes some problems due to
+        // factoring (instead of distributing) constants.
+        if (const Cast *cast = value.as<Cast>()) {
+            if (cast->type.can_represent(cast->value.type()) || cast->type.can_represent(op->type)) {
```

Comment:
Do we really want `can_represent()` here instead of equality? Are there cases where matching to a *specific* type is important?

Reply:
This covers cases like `int32(int16(uint8))` -> `int32(uint8)` or `uint8(uint16(uint32))` -> `uint8(uint32)` It also covers more simple redundant cases like `int32(int32(x))`.